core/src/connection_reuse.rs

```diff
@@ -40,205 +40,556 @@
+            PeerState::Errored(err) => {
+                // Error: return it
+                let io_err = IoError::new(err.kind(), err.to_string());
+                return Err(io_err);
+            }
+            PeerState::Active {
+                muxer,
+                closed,
+                ..
+            } => {
+                if closed.load(Ordering::Acquire) {
+                    (PeerState::Errored(IoError::new(IoErrorKind::BrokenPipe, "Connection closed")),
+                     Err(IoError::new(IoErrorKind::BrokenPipe, "Connection closed")))
+                } else {
+                    // perfect, let's reuse, update the stream_id and
+                    // return a new outbound
+                    trace!(
+                        "Using existing connection to {} to open outbound substream",
+                        addr
+                    );
+                    match poll_for_substream(muxing::outbound_from_ref_and_wrap(muxer.clone()), &addr) {
+                        Ok(None) => return Ok(Async::NotReady),
+                        Ok(Some(stream)) => {
+                            return Ok(Async::Ready((stream, closed.clone())));
+                        }
+                        Err(err) => {
+                            let io_err = IoError::new(err.kind(), err.to_string());
+                            (PeerState::Errored(io_err), Err(err))
+                        }
+                    }
+                }
+            }
+            PeerState::Pending {
+                ref mut future,
+                ref mut notify,
+            } => {
+                // was pending, let's check if that has changed
+                match future.poll() {
+                    Ok(Async::NotReady) => {
+                        // no it still isn't ready, keep the current task
+                        // informed but return with NotReady
+                        notify.insert(TASK_ID.with(|&t| t), task::current());
+                        return Ok(Async::NotReady);
+                    }
+                    Err(err) => {
+                        // well, looks like connecting failed, replace the
+                        // entry then return the Error
+                        trace!("Failed new connection to {}: {:?}", addr, err);
+                        let io_err = IoError::new(err.kind(), err.to_string());
+                        (PeerState::Errored(io_err), Err(err))
+                    }
+                    Ok(Async::Ready((muxer, client_addr))) => {
+                        trace!("Successful new connection to {} ({})", addr, client_addr);
+                        for (_, task) in notify.drain() {
+                            task.notify();
+                        }
+                        let muxer = Arc::new(muxer);
+                        match poll_for_substream(muxing::outbound_from_ref_and_wrap(muxer.clone()), &addr) {
+                            Ok(None) => return Ok(Async::NotReady),
```

Comment:
If we successfully dial a remote but it's not ready to open a substream yet, this means we will poll the dial again the next time and probably panic.